plugins/chat/app/services/base.rb

```diff
@@ -0,0 +1,186 @@
+# frozen_string_literal: true
+
+module Chat
+  module Service
+    module Base
+      extend ActiveSupport::Concern
+
+      class Failure < StandardError
+        attr_reader :context
+
+        def initialize(context = nil)
+          @context = context
+          super
+        end
+      end
+
+      class Contract
+        include ActiveModel::API
+        include ActiveModel::Attributes
+        include ActiveModel::AttributeMethods
+
+        # @!visibility private
+        def self.model_name
+          ActiveModel::Name.new(self, nil, "contract")
+        end
+      end
+
+      class Context < OpenStruct
+        def self.build(context = {})
+          self === context ? context : new(context)
+        end
+
+        def success?
+          !failure?
+        end
+
+        def failure?
+          @failure || false
+        end
+
+        def fail!(context = {})
+          merge(context)
+          @failure = true
+          raise Failure, self
+        end
+
+        def rollback!
+          return false if @rolled_back
+          _called.reverse_each do |service|
+            service.instance_eval(&self.class.rollback_block) if self.class.rollback_block
+          end
+          @rolled_back = true
+        end
+
+        private
+
+        def called!(service)
+          _called << service
+        end
+
+        def _called
+          @called ||= []
+        end
+
+        def merge(other_context = {})
+          other_context.each { |key, value| self[key.to_sym] = value }
+          self
+        end
+      end
+
+      module Helpers
+        def guardian(name, *args)
+          unless context[:guardian].public_send(name, *args)
+            context.fail!("guardian.failed" => name)
```

Comment:
Actually even further, we could just add methods like this to the `Context` to make things nicer:

```ruby
def guardian_failed?
  guardian_failed_error.present?
end

def guardian_failed_error
  result[:guardian_failed]
end

def contract_failed?
  result[:contract_failed]
end
```

We don't want callers to have to dig into things IMO

Reply:
So generally I agree with the `:"xxx"` being ugly and slightly odd. One thing to have in mind is that generally speaking we probably don't want people to interact much with it directly. eg: in controllers we will give a helper to handle all this logic for you.

This being said I wouldn't be against `:guardian_failed`, what do you think @Flink ?

Reply:
Yeah I definitely think the controller can handle this generally, though we do have to keep in mind the Job -> Service and Service -> Service calls, which may need to inspect this manually. IMO the wrapper methods are a nicer way of accessing this since you don't need to know the hash key etc., but will see what Loic thinks too :+1:

Reply:
Yes, having a nice wrapper in controllers is probably what we want/need. This will provide a nice interface to work with. And in the end, a controller should not bother with too many logic details but just know how to handle HTTP stuff based on the service outcome.

Regarding the renaming, why not, the string keys were heavily inspired by Trailblazer. (where you can have things like `contract.default.result`, `policy.custom_name.result`, etc.)